### ISSUE TITLE
question about sample clock frequency

### ISSUE THREAD

**Issue by spzeno** [aMOPel/fomu-trng#1]
Hi, I wonder what is the clock frequency you used during the sampling process. 

Regards.

**Comment by aMOPel** [comment]
It's so long ago that I don't quite know but I can give few pointers:

I think this is where I pass the clk to the neotrng, which would probably be the sample frequency 

https://github.com/aMOPel/fomu-trng/blob/f75e034199c0bcf4a2b894266701ee649f0de263/src/migen/fomu_neoTrng_muacm.py#L157


And I believe this is where that `cd_sys` is actually defined:

https://github.com/litex-hub/litex-boards/blob/041c1607ce7461909cc4076937a55c068e850557/litex_boards/targets/kosagi_fomu.py#L63

Which seems to be 12mhz using the pll to get a quarter of the 48mhz oscillator, iiuc.

But also I use `add_period_constraint` to the `cd_sys` with 24mhz here:

https://github.com/aMOPel/fomu-trng/blob/f75e034199c0bcf4a2b894266701ee649f0de263/src/migen/fomu_neoTrng_muacm.py#L96

I don't remember what that does exactly.

**Comment by spzeno** [comment]
Thanks for your kindly reply! Have you ever tested any other trng implematation that passes the diehard test? I am tring to find a good opensource trng implematation.

**Comment by aMOPel** [comment]
No. Honestly, I struggled to find any implementation that would physically fit on the fomu (I think it was already half full with other stuff I needed) and also an implementation that I could get to build with my limited experience. And I only found a hand full open source trngs to begin with back then iirc. I think I picked this one also because it had parameters to control the size.

The thing I learned during my research was that the randomness performance of the trng on a fpga chip is extremely fragile. 

There are multiple approaches to build a trng but from what I found, only this "free-running ring-oscillators" approach is usable on an fpga. It uses the fact that there are tiny variations in physical signal speed on every wire. It then tries to accumulate this variation into a "random" outcome by letting a signal run many many loops. Anyway the problem is that electrical signals in close proximity on a chip can impact each other. So depending on what other signals are routed near the fro the randomness can be impacted. 

The other problem is that the build tooling tries to fit things onto the board using some algorithm. There are ways to hardcode the position of certain circuits I believe, but I never got that deep into it. So I always let the algorithm determine where to put things including the fro. That means for every configuration that I tested things might well have looked completely different on the board. That's why I tried so many configurations. It was basically a brute force search in a chaotic system.

And then there is still the variation in hardware that comes from the factory. It's possible that on a different fomu chip my results would have been quite different.

Basically to get a meaning full measurement for trng performance you would need to place the fro by hand and also have all the other circuits in fixed positions then do your trng tests on multiple fomu in parallel to compare results.

**Comment by spzeno** [comment]
Thanks a lot for your very kindly reply again! I appreciate all the work you've done and the information you provided!

**Comment by aMOPel** [comment]
Out of curiosity, what are you trying to achieve?

**Comment by spzeno** [comment]
> Out of curiosity, what are you trying to achieve?

I'm tring to impl a hardware ECDSA accelerator, a hardware random number generator is needed as a part of the accelerator.
